Font Library: add progress-bar while uploading font assets

hbhalodia opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
- Add `ProgressBar` component while uploading the font assets.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
- Issue - Closes https://github.com/WordPress/gutenberg/issues/54399

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- This PR adds the progress bar components and hide the drop zone and file upload while assets are being uploaded.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
1. Open site editor.
2. Go to styles and then typography option.
3. Select any style and there would be option for upload.
4. Add the font file there, you would see the progress bar for the moment till file uploads, and we get any notice.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
- None.

## Screenshots or screencast <!-- if applicable -->
- Since the uploading is very fast, you would hardly see the progress bar in action.

[Font Library - Progerss Bar.webm](https://github.com/WordPress/gutenberg/assets/58802366/c2ea716e-e15a-47e3-8284-03e346781794)

hbhalodia committed:
Add progress bar when uploading the font file via site editor

mikachan reviewed (commented):
Thanks for working on this, @hbhalodia!

Currently I'm seeing this in the `isUploading` state:

<img width="958" alt="image" src="https://github.com/WordPress/gutenberg/assets/1645628/429720d1-7bc9-4fa8-8782-d094d6d060ba">

I think it would be good to centre align the progress bar and add some vertical spacing, similar to the mockup from @jasmussen [here](https://github.com/WordPress/gutenberg/issues/54399#issuecomment-1717212256).

hbhalodia commented:
Sure, @mikachan - I would look into this asap.

Thanks.

hbhalodia committed:
Merge branch 'WordPress:trunk' into feat/issue-54399

hbhalodia committed:
Add vertical spacing and make progress bar center align

With this commit, we have centered the progress bar and added some vertical spacing to the progress bar so it look as in center.Also while uploading we have just shown the progress bar and no text before and after, because while uploading it is good to show only progress bar.

hbhalodia commented:
Hi @mikachan, I have revised the code and made the progress bar center aligned and added the vertical spacing. Also in addition to it, I had just showed the progress bar while uploading, since while upload it is good to show just a progress bar and nothing else. Let me know if we need the bottom text while uploading, I can revert that. 

Here is the screenshot how it would be while uploading.

<img width="931" alt="Screenshot 2024-01-04 at 10 22 19 AM" src="https://github.com/WordPress/gutenberg/assets/58802366/58c4ec96-19eb-4c03-a420-e8f56a199c46">

Thank You!

mikachan commented on packages/edit-site/src/components/global-styles/font-library-modal/local-fonts.js:
```diff
@@ -150,29 +156,39 @@ function LocalFonts() {
+						) }
+					/>
+				) }
+				{ isUploading && (
+					<FlexItem>
+						<Spacer margin={ 32 } />
+						<ProgressBar className="font-library-modal__upload-area__progress-bar" />
+						<Spacer margin={ 32 } />
+					</FlexItem>
+				) }
```
I think we can achieve a more consistent layout by re-using the `font-library-modal__upload-area` class, as this already has a fixed height. Something like this:

```suggestion
				{ isUploading && (
					<FlexItem>
						<div className="font-library-modal__upload-area">
							<ProgressBar />
						</div>
					</FlexItem>
				) }
```

We don't want the progress bar to have a grey background, so we could adjust the CSS so the background colour is specific to the upload button:

```
button.font-library-modal__upload-area {
	background-color: #f0f0f0;

}
```

This should create a more consistent layout between the not-uploading is isUploading states:

| Not-uploading | isUploading |
| -------------- | ------------ |
| <img width="977" alt="image" src="https://github.com/WordPress/gutenberg/assets/1645628/f5c133a5-04a5-4779-93c8-861319295d16"> | <img width="977" alt="image" src="https://github.com/WordPress/gutenberg/assets/1645628/47c52b0f-cb37-4418-ae0f-5e9a5409b2d7"> |

hbhalodia replied:
Sure would be looking onto this.

Thanks.

mikachan reviewed (commented):
Thanks @hbhalodia!

> Let me know if we need the bottom text while uploading, I can revert that.

After testing this out, I think we should leave this text in place, as it reduces the visual changes and layout shifts.

Maybe we should also move the Notice component below this text, to further reduce the visual changes, e.g.:

<img width="977" alt="image" src="https://github.com/WordPress/gutenberg/assets/1645628/7ef54899-b1d7-4099-9669-92a3e93ad975">


I've left an inline comment too, as I think we can re-use an existing CSS class rather than the Spacer components. Let me know what you think!

hbhalodia commented:
>I think we should leave this text in place

Sure, Would be keeping as is.

>Maybe we should also move the Notice component below this text.

Yep, we can do it. Would update in the next commit.

Also, would address the inline comment that you have mentioned.

Thanks for the review!

hbhalodia committed:
Add more consistent style for progress bar

Added the consistent style for progress bar and also change the text position before any notice after uploading the fonts in order to reduce the layout shifts.

hbhalodia committed:
Merge branch 'trunk' into feat/issue-54399

hbhalodia commented:
Hi @mikachan, have updated the PR with the required changes, Let me know if that works. Thanks.

| Not Uploading State | Uploading State |
| :-------------: | :-------------: |
| <img width="931" alt="Screenshot 2024-01-04 at 4 59 26 PM" src="https://github.com/WordPress/gutenberg/assets/58802366/c99cf716-ba5c-4607-bf42-fffe358cd4f1"> | <img width="931" alt="Screenshot 2024-01-04 at 5 02 20 PM" src="https://github.com/WordPress/gutenberg/assets/58802366/028dd5fe-68e7-453f-b2b2-f447a214ca93"> |

mikachan reviewed (approved):
Looks great, thanks @hbhalodia! ✨

hbhalodia commented:
Thanks, @mikachan 👍.
